Migration: Fixing import  issues / Speed up repeated unknown pod detection

tclaus opened the pull request:
It handles some migration issues listed in #8256, please check @cmrd-senya, @jhass, @Flaburgan .
(Import stops when posts comments or likes don't have a parent/author),
It handles #8019 (caching person fetches from remotes)

For this reason I changed the Person.find_or_fetch_by_identifier function. It now won't raise a NotFetchable exception, it just returns nil or the person object. Some callers rely on nil (which was never returned), some other rely on exception. IMHO this function should return an existing person from local store or by fetching, but never raise an exception. 
Now,  Person.find_or_fetch_by_identifier  returns a person, or nil. Callers are adapted to this behaviour.

@cmrd-senya : I left over a failing test, which I can't fix, maybe you have on eye on this? 
@jhass : can you check import - times with your archive file?

diaspora-code-review commented on spec/federation_callbacks_spec.rb:
```diff
@@ -231,14 +231,12 @@
       diaspora_id = Fabricate.sequence(:diaspora_id)
       expect(Person).to receive(:find_or_fetch_by_identifier).with(diaspora_id)
-        .and_raise(DiasporaFederation::Discovery::DiscoveryError)
+        .and_return(nil)
```
Layout/MultilineMethodCallIndentation: Align `.and_return` with `.with` on line 236.

tclaus commented:
Reminder: To enable/disable local caching on your dev machine you can toggle it (without changing configuration files) with: 
$: rails dev:cache

I also added a statement to the spec helpers to clear any caches between each test.

diaspora-code-review commented on spec/lib/archive_importer/post_importer_spec.rb:
```diff
@@ -113,5 +113,39 @@
+    end
   end
-end
+end
```
Layout/TrailingEmptyLines: Final newline missing.

SuperTux88 commented on lib/archive_importer/own_relayable_importer.rb:
```diff
@@ -19,7 +21,9 @@ def fetch_parent(data)
       }
       entity = Diaspora::Federation::Mappings.model_class_for(type).find_by(guid: data.fetch(:parent_guid))
-      data[:parent] = Diaspora::Federation::Entities.related_entity(entity)
+      raise NoParentError if entity.nil?
```
The same question as in https://github.com/diaspora/diaspora/pull/8010#discussion_r281341950 still exists. In theory this shouldn't happen, because in the validate step it already tries to fetch all the entities here

https://github.com/diaspora/diaspora/blob/f85f167f505985c6c34dc77f5c29493e28d7aa20/lib/archive_validator/own_relayable_validator.rb#L15-L17

using

https://github.com/diaspora/diaspora/blob/f85f167f505985c6c34dc77f5c29493e28d7aa20/lib/diaspora/federated/fetchable.rb#L9-L17

which already returns nil when not being able to be fetched, and if the parent is nil, it's not valid:

https://github.com/diaspora/diaspora/blob/a3196a108946eb331d007add064182e558e448aa/lib/archive_validator/relayable_validator.rb#L16-L18

which then should remove all invalid entities during the validation phase, so they aren't tried to be actually imported:

https://github.com/diaspora/diaspora/blob/f85f167f505985c6c34dc77f5c29493e28d7aa20/lib/archive_validator/collection_validator.rb#L9-L13

If this doesn't work as expected, it also means a lot of the things are done twice, which isn't intended (and could also lead to longer runtimes). If you have steps to reproduce this and can debug why entities that don't exist still end up at this point and aren't filtered out earlier already, it would be good to know and should be fixed there instead.

tclaus replied:
The part: find_by(guid: data.fetch(:**parent_guid**))  makes the problem here, not detected by the validator. 
But we already talked that it seems to be inconvenient to fetch everything in the validation part.

SuperTux88 reviewed (changes requested):
This isn't a full review yet, just a few comments about some basic stuff which probably changes a lot of the rest of the PR anyway.

tclaus commented:
@SuperTux88 I give up. Sorry - I didn't understand the validation process in depth so that I can check all needed authors, parents and entries and fix tests and implement all as a validation.

tclaus committed:
Fixing migration import errors
Fixing raising NotFetchable error
Fixing not found a parent on entities

Flaburgan commented:
> @SuperTux88 I give up. Sorry - I didn't understand the validation process in depth so that I can check all needed authors, parents and entries and fix tests and implement all as a validation.

If I understand correctly, you since had a call with @SuperTux88 to solve this point. So what is the status of that pull request now?

tclaus commented:
My status is, that first idea was to push checks into the validation part, which did not succeed because of needed code changes. Another (more inportant) point was that fetching everything during validation is not what a validation process should do. 

After some code attempts, I implemented the smallest possible solution in this PR. 
To get forward and have one working branch I also rebased the upload-migration file branch  to this work. 

Status is: it just works. (Until someone else Codes a „better“ solution). 



> Am 14.09.2021 um 22:15 schrieb Fla ***@***.***>:
> 
> ﻿
> @SuperTux88 I give up. Sorry - I didn't understand the validation process in depth so that I can check all needed authors, parents and entries and fix tests and implement all as a validation.
> 
> If I understand correctly, you since had a call with @SuperTux88 to solve this point. So what is the status of that pull request now?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

tclaus commented:
@SuperTux88 As discussed, I have tried to make the smallest possible working solution without rebuilding the validation part. This solution gives what it should give.
With this in mind - could you check this again and possible approve (and merge) this? 
It will be let the #8274 working.

SuperTux88 commented:
I ended up merging https://github.com/diaspora/diaspora/pull/8010 because it had separate commits for the two different problems and also had a test for both problems. But I merged in the test from here since this was more generic (tests generic fetch problem and not only discovery errors).

Thanks for the work on this anyway since we finally found an answer for the original question in that PR :cookie:

Flaburgan commented:
Awesome to finally see this merged!

tclaus commented:
Thanks for your analysis - to finally found a solution.
